File: myExpenses/src/main/java/org/totschnig/myexpenses/model/CrStatus.java
package org.totschnig.myexpenses.model;

import android.graphics.Color;

import org.totschnig.myexpenses.R;
import org.totschnig.myexpenses.util.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.StringRes;

public enum CrStatus {
  UNRECONCILED(Color.GRAY, ""), CLEARED(Color.BLUE, "*"), RECONCILED(Color.GREEN, "X"), VOID(Color.RED, "V");
  public int color;
  @NonNull
  public String symbol;

  CrStatus(int color, @NonNull String symbol) {
    this.color = color;
    this.symbol = symbol;
  }

  public static final String JOIN;

  static {
    JOIN = TextUtils.joinEnum(CrStatus.class);
  }

  public static CrStatus fromQifName(String qifName) {
    if (qifName == null)
      return UNRECONCILED;
    if (qifName.equals("*") || qifName.equalsIgnoreCase("C")) {
      return CLEARED;
    } else if (qifName.equalsIgnoreCase("X") || qifName.equalsIgnoreCase("R")) {
      return RECONCILED;
    } else {
      return UNRECONCILED;
    }
  }

  @StringRes
  public int toStringRes() {
    switch (this) {
      case CLEARED:
        return R.string.status_cleared;
      case RECONCILED:
        return R.string.status_reconciled;
      case UNRECONCILED:
        return R.string.status_uncreconciled;
      case VOID:
        return R.string.status_void;
    }
    return 0;
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/model/CrStatus.kt
package org.totschnig.myexpenses.model

import android.content.Context
import androidx.annotation.ColorRes
import androidx.annotation.StringRes
import androidx.core.content.ContextCompat
import com.google.android.material.color.MaterialColors
import org.totschnig.myexpenses.R
import org.totschnig.myexpenses.util.TextUtils.joinEnum
import java.util.Locale

enum class CrStatus(@ColorRes val color: Int, val symbol: Char?) {
    UNRECONCILED(R.color.UNRECONCILED, null),
    CLEARED(R.color.CLEARED, '*'),
    RECONCILED(R.color.RECONCILED, 'X'),
    VOID(R.color.VOID, 'V');

    @StringRes
    fun toStringRes() = when (this) {
        CLEARED -> R.string.status_cleared
        RECONCILED -> R.string.status_reconciled
        UNRECONCILED -> R.string.status_unreconciled
        VOID -> R.string.status_void
    }

    fun toColorRoles(context: Context) = MaterialColors.getColorRoles(
        context,
        ContextCompat.getColor(context, color)
    )

    companion object {
        @JvmField
        val JOIN: String = joinEnum(CrStatus::class.java)

        @JvmStatic
        fun fromQifName(qifName: String?) =
            when(qifName?.uppercase(Locale.getDefault())) {
                "*", "C" -> CLEARED
                "X", "R" -> RECONCILED
                "V" -> VOID
                else -> UNRECONCILED
            }

        val editableStatuses = arrayOf(UNRECONCILED, CLEARED, VOID)
    }
}